### DISCUSSION TITLE
Thoughts about JSON

### DISCUSSION THREAD

**Discussion by mosheduminer** [titzer/virgil#115] [General]
I've been thinking about the prerequisites to JSON (de)serialization to/from data structures in virgil. What I would like is a library like serde and serde_json in rust, where data can be deserialized directly into a struct, and serialized directly from a struct. 

However, it seems to me that I'm virgil the only way to deserialize from json would be to deserialize into a recursive ADT of hashmaps / arrays / allowed primitives. If I wanted to get the data into (or out of) a well defined data structure though, i would to write additional code to connect the ADT structure to a class (or any other virgil structure).

What this leads me to, is the fact that there is no reflection in virgil, even in compile time code. If it were possible to allow reflection at compile time, I imagine it would open up static metaprogramming such as is necessary for a library like serde. However, I know that reflection is listed as something that virgil lacks, and I would like to know if the reasons for that (I assume performance? Are there also other considerations) also preclude compile time reflection, and if not, whether you would consider adding it to your roadmap. I understand that virgil is in a sense an expiriment (?), and I of course do not wish to imply that there is any onus on you to add a particular feature.

**Comment by k-sareen** [comment]
I think you'll find that the proposed WIP `struct` data structures precisely do what you want -- having a well defined data structure over an array of bytes essentially. It's supposed to be used for low-level code such as networking and interacting with the kernel, and it could be used for JSON parsing as well.

@titzer can correct me if I'm wrong, but I feel like it should be possible with `struct`s (assuming I understand them correctly)

EDIT: The above is based on the assumption that you know the structure of the JSON data, If you're trying to make a general library for it, then I think `struct`s may not be enough? I'm not sure.

**Reply by mosheduminer** [reply]
I'm sorry, but why) I don't understand why structs would be better for this purpose? My problem is not wanting to write custom code for each data structure I want to convert to/from json, a problem which would maybe be solved with appropriate compile time features. But I don't see why structs would be better for this?

**Reply by k-sareen** [reply]
See comment below; I am/was mistaken -- it may not work unless you have a fixed size.

**Comment by k-sareen** [comment]
Actually on second thought -- `struct`s may not be sufficient as they assume a fixed size? I'm a bit hazy on the capabilities of `struct`s, but if they are fixed size only, then a string format like JSON, where each instance of the same JSON structure may have different lengths, may not work.

**Comment by titzer** [comment]
> I've been thinking about the prerequisites to JSON (de)serialization to/from data structures in virgil. What I would like is a library like serde and serde_json in rust, where data can be deserialized directly into a struct, and serialized directly from a struct.
> 
> However, it seems to me that I'm virgil the only way to deserialize from json would be to deserialize into a recursive ADT of hashmaps / arrays / allowed primitives. If I wanted to get the data into (or out of) a well defined data structure though, i would to write additional code to connect the ADT structure to a class (or any other virgil structure).
> 
> What this leads me to, is the fact that there is no reflection in virgil, even in compile time code. If it were possible to allow reflection at compile time, I imagine it would open up static metaprogramming such as is necessary for a library like serde. However, I know that reflection is listed as something that virgil lacks, and I would like to know if the reasons for that (I assume performance? Are there also other considerations) also preclude compile time reflection, and if not, whether you would consider adding it to your roadmap. I understand that virgil is in a sense an expiriment (?), and I of course do not wish to imply that there is any onus on you to add a particular feature.

The motivation for not (yet) having reflection has been primarily about memory usage and optimizability of Virgil code. Reflection requires having source-level metadata connected to program values, and that metadata takes a lot of space. It can also prevent a lot of the static analysis the compiler does to optimize and remove dead code.

That said, reflection only at compile-time would restrict those problems to a phase that can be completed, after which optimization can happen. That's something I could see as a future feature.

Failing that, there are a number of non-reflective things that you can try first.

1. Use dynamically typed (or "stringly"-typed) objects to represent JSON, so that you can write code like:

```
type JSONVal {
 ...
 def toObj() -> JSONVal.Obj;
 def toInt() -> int;
}
var x = deserializeJSON(input);
var y = x["field1"].toObj();
var z = y["field2"].toInt();
```

2. Minimize the amount of manual code you'd have to write by designing a serializer combinator library:

```
class MyClass(var x: int, var y: int) {
}
// deserializes directly into constructor of MyClass
def deserializeMyClass = Deserializer.new2<MyClass>(MyClass.new, Deserializer.i32("x"), Deserializer.i32("y"));
// serializes using MyClass.field accessor functions
def serializeMyClass = Serializer.new2<MyClass>(Serializer.i32(MyClass.x), Serializer.i32(MyClass.y));
```

The above is a bit verbose, but you could:

3. Use the Virgil parser (from src/aeneas) as a library in another tool to process Virgil source code and generate more Virgil source that serializes/deserializes a type.

**Reply by mosheduminer** [reply]
I'm working on an implementation, hopefully I'll have something to share next week. The API surface will be a variation on 2.

**Reply by titzer** [reply]
Cool, happy to accept a PR. It would live nicely under `lib/json`.

**Reply by mosheduminer** [reply]
> Cool, happy to accept a PR. It would live nicely under `lib/json`.

Sure. That would be cool!

**Reply by mosheduminer** [reply]
My current work is at https://gist.github.com/mosheduminer/996a6016642cf0c5577e92428dc75520. It includes a small parser combinator library, a JSON parser using the combinator library, and a transformer that transforms the parser result to an ADT of the form
```
type JsonAst {
	case Null;
	case String(value: string);
	case Number(value: string);
	case Bool(value: bool);
	case Array(array: Array<JsonAst>);
	case Object(object: Array<(string, JsonAst)>);
}
```
Also, the ADT has `toString()` methods, allowing serialization from the AST to a string.

Next up to make it easier to deserialize the AST to a specific data structure.

**Reply by titzer** [reply]
Looks really cool! 

I had a quick scan through your code and it looks very readable. 
We could even factor out the parse combinator into `lib/parse` and clean it up for others to use as well.

**Reply by mosheduminer** [reply]
I updated the gist, the API to create a deserializer is roughly:
```
def deserializer = deArray(
	deObject2(
		JsonObjectDeserializedExample.new,
		("a", deInt32),
		("b", deInt32)
	)
);
```
I'm not sure I like it though.

**Reply by mosheduminer** [reply]
I'm gonna rewrite the API before making a PR.

**Reply by titzer** [reply]
sgtm

**Reply by mosheduminer** [reply]
Hey, I just updated the gist, do you mind taking a look at the TODOs there? Mainly I want to know about parsing floats (we don't have that yet, right?) and about methods on ADTs that is not working as I would expect.

Separately, is there anything more I need for a PR? Will I need tests etc. (I don't know how the test system in this repo works)? Or should I just open a PR with a file in a `lib/parse` directory and another in `lib/json`?

**Reply by k-sareen** [reply]
> https://gist.github.com/mosheduminer/996a6016642cf0c5577e92428dc75520#file-json-v3-L90-L93

You could force a cast instead of re-wrapping the number. Something like `JsonValue.Number.!(json).i32()`. You could also do a 

```
match (json) {
  x: Number(v) => { }
}
```
and then use `x` inside the case.

But yeah, it would be nice if the type system would know that `json` is a `Number` at that point.

**Reply by mosheduminer** [reply]
Doing this
```
match (json) {
  x: Number(v) => { }
}
```
seems to result in a parse error for me?

**Reply by k-sareen** [reply]
Oh... Maybe that syntax is only if you don't destructure the case? It's been a while, but I'm fairly certain that `x: Case` is valid syntax.

Edit: Yeah see https://github.com/titzer/virgil/blob/master/aeneas/src/mach/MachBackend.v3#L672-L675. It may be that you can't destructure the case.

**Reply by mosheduminer** [reply]
I'll try tomorrow morning then... It's late where I am.

**Reply by mosheduminer** [reply]
Worst case I can cast, as you suggested.

**Reply by k-sareen** [reply]
For what it's worth, the following does work:

```
type A {
        case B(v: int);
        case C(c: long);
}

def main() {
        def s = A.B(10);
        match (s) {
                x: A.B => {
                        System.puti(x.v);
                        System.puts("\n");
                }
                _ => {
                        System.error("Error!", "Unexpected case");
                }
        }
}
```

**Reply by mosheduminer** [reply]
I see, so interestingly, only this `x: JsonValue.Number` works for me:
```
match (json) {
	x: JsonValue.Number => return x.i32();
	_ => return Result.Err("Expected to find Number");
}
```
but `x: Number` and `x: JsonValue.Number(v)` do not. I would expect that at least `x: Number` should work though?

**Comment by titzer** [comment]
The syntax for `match` is either a new variable binding with a type or a destructure:

```
type T {
  case X(a: int) { }
  case Y(b: int) { }
}

def m(v: T) {
  match (v) { // query, cast, bind new variable
    x: T.X => { }
    x: T.Y => { }
  }
  match (v) { // destructure
    X(a) => { }
    Y(b) => { }
  }
}
```

There are routines to parse floats inside the compiler (look for `NumberParser.v3` and `FloatSketch.v3`). They have just not been pulled out to a library yet.

**Reply by mosheduminer** [reply]
Just to let you know where I'm holding - I have everything down except handling floats and better handling of nullable values in the when deserializing to objects. I haven't worked on it in a bit, I'll try to get back to it and implement float parsing (and maybe a way to make things nullable), then make a PR.

**Reply by mosheduminer** [reply]
@titzer hey, I know it's been a long while, but I'd like to get working on this again and finish it. Last time, the primary blocker was that I felt unsure extracting the float parsing logic from the NumberParser in aeneas. Do you mind please extracting it into it the utils?

**Reply by titzer** [reply]
Hello again! What about making a copy and adapting it to your needs? We can combine them again after seeing how it fits into the JSON utility.

**Reply by mosheduminer** [reply]
Alright, I'll try that.

**Reply by mosheduminer** [reply]
The `NumberParser` and its methods seem to rely on a number of types and classes internal to aeneas (such as `Type`, `Val`, `Int`, `Float` and `FpSketch`), so I'm having difficulty copying/using it.

**Reply by titzer** [reply]
Oh, I see. Some of those things can be extracted out more easily than others. For example I think `FpSketch` shouldn't depend on the `Type` and `Val` too much? I think you'll have to stub out `Type` to be an enum; there are only a small set of types that a number in Virgil can have.

**Reply by mosheduminer** [reply]
@titzer I finally extracted FpSketch and a few classes from other files (and replaced Type) to get it to work. But I'm not sure how to convert the `NumberParser.val` to a float or double, I tried [this](https://github.com/mosheduminer/virgil-parser-combinators/blob/master/Json.v3#L50) and it gives a `!TypeCheckException`, so clearly it isn't the right way to do things. Any suggestions?

**Reply by titzer** [reply]
You'll want to use `float.view` or `double.view`, which reinterpret the bits (i.e. no rounding or truncating, just a bit-level cast).

**Reply by mosheduminer** [reply]
Thanks! It works now 🥳.

```sh
~/virgil/parser-combinators$ make test
rm -f Test
v3c-x86-64-linux Json.v3 PC.v3 Test.v3 ../virgil/lib/util/*.v3 ../virgil/lib/test/*.v3 ./extracted/*.v3
~/virgil/parser-combinators$ ./Test
##>3 unit tests
##+parsing
##-ok
##+deserializing
##-ok
##+fail_deserializing
##-ok
```

**Reply by mosheduminer** [reply]
I remember at some point in the past (can't find it now) you suggested that the parser combinator (and the JSON parser too?) be contributed to this repo as a PR. Are you still interested, and if so do you have any guidelines / suggestions for that?

**Reply by titzer** [reply]
Nice!

Let's do the JSON parsing utility as a first PR, then we can design parser combinators as a follow-up. Wdyt?

**Reply by mosheduminer** [reply]
Well, my JSON parsing is implemented on top of my parser combinator system, so I don't think that ordering will work 😃.

**Reply by mosheduminer** [reply]
If you prefer an independent JSON parser I can maybe work on that though.
